File: spec/cpp_stl_11/test_switch_manual_int_size.cpp
#include <boost/test/unit_test.hpp>

#include <switch_manual_int_size.h>

#include <iostream>
#include <fstream>
#include <vector>

#include "helpers.h"

BOOST_AUTO_TEST_CASE(test_switch_manual_int_size) {
    std::ifstream ifs("src/switch_tlv.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    switch_manual_int_size_t* r = new switch_manual_int_size_t(&ks);

    BOOST_CHECK_EQUAL(r->chunks()->size(), 4);

    BOOST_CHECK_EQUAL(r->chunks()->at(0)->code(), 0x11);
    switch_manual_int_size_t::chunk_t::chunk_meta_t* meta =
        static_cast<switch_manual_int_size_t::chunk_t::chunk_meta_t*>(r->chunks()->at(0)->body());
    BOOST_CHECK_EQUAL(meta->title(), "Stuff");
    BOOST_CHECK_EQUAL(meta->author(), "Me");

    BOOST_CHECK_EQUAL(r->chunks()->at(1)->code(), 0x22);
    switch_manual_int_size_t::chunk_t::chunk_dir_t* dir =
        static_cast<switch_manual_int_size_t::chunk_t::chunk_dir_t*>(r->chunks()->at(1)->body());
    COMPARE_ARRAY(std::string, dir->entries(), "AAAA", "BBBB", "CCCC");

    BOOST_CHECK_EQUAL(r->chunks()->at(2)->code(), 0x33);
    std::string str1 = r->chunks()->at(2)->_raw_body();
    BOOST_CHECK_EQUAL(str1, std::string("\x10\x20\x30\x40\x50\x60\x70\x80"));

    BOOST_CHECK_EQUAL(r->chunks()->at(3)->code(), 0xff);
    std::string str2 = r->chunks()->at(3)->_raw_body();
    BOOST_CHECK_EQUAL(str2, std::string(""));

    delete r;
}

File: spec/rust/tests/test_switch_manual_int_size.rs
#![allow(dead_code)]
use std::fs;

extern crate kaitai;
use self::kaitai::*;
mod formats;
use formats::switch_manual_int_size::*;

#[test]
fn test_switch_manual_int_size() {
    let bytes = fs::read("../../src/switch_tlv.bin").unwrap();
    let reader = BytesReader::new(&bytes);
    let mut r = SwitchManualIntSize::default();

    if let Err(err) = r.read(&reader, None, Some(KStructUnit::parent_stack())) {

        panic!("{:?}", err);
    }
    assert_eq!(4, r.chunks().len());

    assert_eq!(17, r.chunks()[0].code);
    if let SwitchManualIntSize_Chunk_Body::SwitchManualIntSize_Chunk_ChunkMeta(s) =  r.chunks[0].body() {
        assert_eq!("Stuff", s.title());
        assert_eq!("Me", s.author());
    } else {
        panic!("expected enum SwitchManualIntSize_Chunk_ChunkMeta");
    }

    assert_eq!(34, r.chunks()[1].code);
    if let SwitchManualIntSize_Chunk_Body::SwitchManualIntSize_Chunk_ChunkDir(s) =  r.chunks[1].body() {
        let strings : Vec<String> = vec!["AAAA", "BBBB", "CCCC"].iter().map(|&s| s.to_string() ).collect();
        assert_eq!(strings, *s.entries());
    } else {
        panic!("expected enum SwitchManualIntSize_Chunk_ChunkDir");
    }
 
    assert_eq!(51, r.chunks()[2].code);
    if let SwitchManualIntSize_Chunk_Body::Bytes(s) =  r.chunks[2].body() {
        let raw : Vec<u8> = vec![0x10, 0x20, 0x30, 0x40, 0x50, 0x60, 0x70, 0x80];
        assert_eq!(raw, *s);
    } else {
        panic!("expected enum Bytes");
    }
 
    assert_eq!(255, r.chunks()[3].code);
    if let SwitchManualIntSize_Chunk_Body::Bytes(s) =  r.chunks[3].body() {
        let raw : Vec<u8> = vec![];
        assert_eq!(raw, *s);
    } else {
        panic!("expected enum Bytes");
    }
}
